sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdaterFactory.java

```diff
@@ -0,0 +1,979 @@
+        WritableColumnVector values,
+        WritableColumnVector dictionaryIds,
+        Dictionary dictionary) {
+      values.putByte(offset, (byte) dictionary.decodeToInt(dictionaryIds.getDictId(offset)));
+    }
+  }
+
+  private static class ShortUpdater implements ParquetVectorUpdater {
+    @Override
+    public void updateBatch(
+        int total,
+        int offset,
+        WritableColumnVector values,
+        VectorizedValuesReader valuesReader) {
+      valuesReader.readShorts(total, values, offset);
+    }
+
+    @Override
+    public void update(
+        int offset,
+        WritableColumnVector values,
+        VectorizedValuesReader valuesReader) {
+      values.putShort(offset, valuesReader.readShort());
+    }
+
+    @Override
+    public void decodeSingleDictionaryId(
+        int offset,
+        WritableColumnVector values,
+        WritableColumnVector dictionaryIds,
+        Dictionary dictionary) {
+      values.putShort(offset, (short) dictionary.decodeToInt(dictionaryIds.getDictId(offset)));
+    }
+  }
+
+  private static class IntegerWithRebaseUpdater implements ParquetVectorUpdater {
+    private final boolean failIfRebase;
+
+    IntegerWithRebaseUpdater(boolean failIfRebase) {
+      this.failIfRebase = failIfRebase;
+    }
+
+    @Override
+    public void updateBatch(
+        int total,
+        int offset,
+        WritableColumnVector values,
+        VectorizedValuesReader valuesReader) {
+      valuesReader.readIntegersWithRebase(total, values, offset, failIfRebase);
+    }
+
+    @Override
+    public void update(
+        int offset,
+        WritableColumnVector values,
+        VectorizedValuesReader valuesReader) {
+      int julianDays = valuesReader.readInteger();
+      values.putInt(offset, rebaseDays(julianDays, failIfRebase));
+    }
+
+    @Override
+    public void decodeSingleDictionaryId(
+        int offset,
+        WritableColumnVector values,
+        WritableColumnVector dictionaryIds,
+        Dictionary dictionary) {
+      int julianDays = dictionary.decodeToInt(dictionaryIds.getDictId(offset));
+      values.putInt(offset, rebaseDays(julianDays, failIfRebase));
+    }
+  }
+
+  private static class LongUpdater implements ParquetVectorUpdater {
+    private final boolean downCastLongToInt;
+
+    LongUpdater(boolean downCastLongToInt) {
+      this.downCastLongToInt = downCastLongToInt;
+    }
+
+    @Override
+    public void updateBatch(
+        int total,
+        int offset,
+        WritableColumnVector values,
+        VectorizedValuesReader valuesReader) {
+      if (downCastLongToInt) {
+        for (int i = 0; i < total; ++i) {
+          values.putInt(offset + i, (int) valuesReader.readLong());
+        }
+      } else {
+        valuesReader.readLongs(total, values, offset);
+      }
+    }
+
+    @Override
+    public void update(
+        int offset,
+        WritableColumnVector values,
+        VectorizedValuesReader valuesReader) {
+      if (downCastLongToInt) {
```

Comment:
I'm wondering if we should have 2 updater implementations. Here we check this if condition for every function call.

Reply:
Yea I thought about it too. I think it's a good idea. Will do.